src/terminal/adapter/ut_adapter/adapterTest.cpp

```diff
@@ -1286,42 +1286,42 @@ class AdapterTest
         case DispatchTypes::GraphicsOptions::ForegroundBlack:
             Log::Comment(L"Testing graphics 'Foreground Color Black'");
             _testGetSet->_attribute = FOREGROUND_RED | FOREGROUND_GREEN | FOREGROUND_BLUE | FOREGROUND_INTENSITY;
-            _testGetSet->_expectedAttribute = 0;
+            _testGetSet->_expectedAttribute.SetIndexedForeground(0);

```

Comment:
This one stands out to me as a change in test behavior -- on the left side it's replacing and on the right side it's additive, right?

This might not be a _meaningful_ change, of course, I just can't tell from here

Reply:
Yeah, this is relying on the fact that the `_expectedAttribute` is initialized in `PrepData`. I was trying to keep the changes minimal, but it would probably be clearer what's going on if I set `_expectedAttribute` to `_attribute` before updating the foreground index, the same way I've done the background tests below this. Would that make sense to you?

Reply:
I’m fine with this honestly. Just so long as we understand where we’re changing the behavior and that it’s not impactful.

Idly, I wonder if our internal APIs should eventually move towards immutable structs and “WithXxx” functions that return a copy. That’s not work for this PR, but maybe good for me to file as a CodeHealth issue. Dunno. It’s germane here because it would become `_attribute.WithIndexedFG()`

Reply:
That's an interesting thought. It would certainly make these test more readable, but I'd be curious to see how well something like that optimises and how well the pattern works in the main code. I wouldn't want to add those methods exclusively for use by the unit tests, because that means the tests aren't actually testing the real code - they'd just be testing their own private code path.

Reply:
Just for fun, I threw something silly into godbolt to see how it optimizes.

https://godbolt.org/z/UZik2K

```asm
void usingWith(void) PROC                              ; usingWith, COMDAT
$LN6:
        sub     rsp, 40                             ; 00000028H
        call    QWORD PTR __imp_TextAttribute CantOptimizeMe_Sucker(void)
        movzx   ecx, al
        shr     ax, 8
        or      cl, 16
        mov     BYTE PTR $T1[rsp+1], al
        mov     BYTE PTR $T1[rsp], cl
        lea     rcx, QWORD PTR newAttribute$[rsp]
        movzx   eax, WORD PTR $T1[rsp]
        mov     WORD PTR newAttribute$[rsp], ax
        call    QWORD PTR __imp_void CantOptimizeThis_Either(TextAttribute const &)
        add     rsp, 40                             ; 00000028H
        ret     0
void usingWith(void) ENDP 
```

It does end up a little more verbose than the alternative (setting it directly):


```asm
void usingSetter(void) PROC                                ; usingSetter, COMDAT
$LN6:
        sub     rsp, 40                             ; 00000028H
        call    QWORD PTR __imp_TextAttribute CantOptimizeMe_Sucker(void)
        mov     WORD PTR secondOldAttr$[rsp], ax
        lea     rcx, QWORD PTR secondOldAttr$[rsp]
        or      BYTE PTR secondOldAttr$[rsp], 16
        call    QWORD PTR __imp_void CantOptimizeThis_Either(TextAttribute const &)
        add     rsp, 40                             ; 00000028H
        ret     0
void usingSetter(void) ENDP  
```

Inconclusive, and probably micro-optimization, but we also don't need to do the With() work now. :grin:

Reply:
clang optimizes them to the same thing. LOL.

Reply:
Note that the `With...` version does tend to get a worse the more fields you add though. Well except in clang, which is clearly using some kind of alien technology. 👽

Anyway, for now I've checked in a commit that initialises `_expectedAttribute` to `_attribute` before calling `SetIndexedForeground`. I think that makes things a little clearer, and is also more consistent with the `SetIndexedBackground` test cases.